### PR TITLE
Add blur, shadow, and corner radius to layer-shell surfaces

### PR DESCRIPTION
**Pull request by ErikReider**
Adds border radius and blur to layer-shell surfaces (not background and bottom layers).

Requires #157

Todo:
- [x] Command to add layer namespace with effects
- [x] Handle blur
- [x] Handle shadow
- [x] Handle corner radius

- [x] Add layer-effect checking in `get_workspace_effect_info`
- [x] Change config command syntax to match `for_window`
- [x] Add method of viewing all layer namespaces in sway-ipc
- [x] Make the command syntax similar to `for_window`
- [x] Code cleanup (change the config syntax?)
- [x] Disable blur xray for bottom layer surfaces

image

Example command:
`layer_effects "waybar" blur enable; shadows enable; corner_radius 6`

### CODE REVIEW

**Comment by ErikReider** [comment]
I skipped adding shadows due to them only being rendered for the base surface. Would be easier if we render shadows in `render_surface_iterator`. Shadows could be moved into there in this commit if desired :)

**Comment by ErikReider** [comment]
Seems like moving the shadows into `render_surface_iterator` fixes #141 :)

File: README.md [hunk 1]
```diff
@@ -31,6 +31,8 @@ Sway is an incredible window manager, and certainly one of the most well establi
     - `blur_xray enable|disable`
     - `blur_passes <integer value 0 - 10>`
     - `blur_radius <integer value 0 - 10>`
++ LayerShell effects: *ONLY ON SWAYFX-GIT, NOT YET RELEASED*
+    - `layer_effects <layer namespace, eg: "waybar"> blur shadow corner_radius`
```
Comment (WillPower3309) [inline README.md:35]: not sure how, but we should indicate that a user can include any of the options between `blur`, `shadow`, and `corner_radius`

File: sway/desktop/output.c [hunk 2]
```diff
@@ -199,6 +199,14 @@ void output_layer_for_each_toplevel_surface(struct sway_output *output,
 	wl_list_for_each(layer_surface, layer_surfaces, link) {
 		struct wlr_layer_surface_v1 *wlr_layer_surface_v1 =
 			layer_surface->layer_surface;
+		struct fx_render_data *data = user_data;
+		struct layer_effects *effects = layer_surface->effects;
+		if (effects) {
+			data->deco_data.blur = effects->blur && config->blur_enabled;
+			data->deco_data.shadow = effects->shadow && config->shadow_enabled;
+			data->deco_data.corner_radius = effects->corner_rounding? config->corner_radius: 0;
```
Comment (WillPower3309) [inline sway/desktop/output.c:207]: I think we can discuss omitting the `&& config->EFFECT` checks here. Seems weird to have a bar not be blurred if its specified in the layer config but disabled globally.

File: sway/desktop/output.c [hunk 3]
```diff
@@ -199,6 +199,14 @@ void output_layer_for_each_toplevel_surface(struct sway_output *output,
 	wl_list_for_each(layer_surface, layer_surfaces, link) {
 		struct wlr_layer_surface_v1 *wlr_layer_surface_v1 =
 			layer_surface->layer_surface;
+		struct fx_render_data *data = user_data;
+		struct layer_effects *effects = layer_surface->effects;
+		if (effects) {
```
Comment (WillPower3309) [inline sway/desktop/output.c:204]: Would the effects struct always be defined? Could remove this check

File: sway/desktop/render.c [hunk 4]
```diff
@@ -364,7 +408,16 @@ static void render_surface_iterator(struct sway_output *output,
 	deco_data.corner_radius *= wlr_output->scale;
 
 	// render blur (view->surface == surface excludes blurring subsurfaces)
-	if (deco_data.blur && should_parameters_blur() && view->surface == surface) {
+	bool is_subsurface = true;
+	bool is_floating = true;
+	if (data->is_toplevel_surface) {
+		is_subsurface = false;
+		is_floating = !config->blur_xray;
```
Comment (WillPower3309) [inline sway/desktop/render.c:415]: lets rename `is_floating` to `should_optimize_blur` or something more descriptive

File: sway/desktop/output.c (hunk 3 above)
Comment (ErikReider) [reply sway/desktop/output.c:204]: Nah, it's a pointer and only gets defined when a suitable `layer_effects` config is found

File: sway/desktop/render.c (hunk 4 above)
Comment (ErikReider) [reply sway/desktop/render.c]: Fixed by: [7355f7c](https://github.com/WillPower3309/swayfx/pull/144/commits/7355f7c4e7951fc788b2891fb35087add3eafdc8)

File: README.md (hunk 1 above)
Comment (ErikReider) [reply README.md:35]: Would something like this be better?:

```markdown
+ LayerShell effects: *ONLY ON SWAYFX-GIT, NOT YET RELEASED*
    - `layer_effects <layer namespace, eg: "waybar"> <effects, eg: blur shadow corner_radius>`
```

**Comment by ErikReider** [comment]
I updated the description with more discussion topics :)

File: sway/ipc-json.c [hunk 5]
```diff
@@ -331,6 +333,66 @@ static void ipc_json_describe_enabled_output(struct sway_output *output,
...
+				case ZWLR_LAYER_SHELL_V1_LAYER_BOTTOM:
+					layer_name = "bottom";
+					break;
+				case ZWLR_LAYER_SHELL_V1_LAYER_TOP:
+					layer_name = "top";
+					break;
+				case ZWLR_LAYER_SHELL_V1_LAYER_OVERLAY:
+					layer_name = "overlay";
+					break;
+			}
```
Comment (WillPower3309) [inline sway/ipc-json.c:360]: Need to fix a maybe uninitializaed warning here (I can do this, just commenting for my own reminder)

**Comment by WillPower3309** [comment]
Moving to render_surface_iterator breaks the smart_corner_radius shadow fixes:  https://github.com/WillPower3309/swayfx/commit/74e7012b21d48200cf67608f965befe5a308a8bb

I'll extract the render_surface_iterator changes to a different PR to help focus on this piece by piece and ensure nothing breaks along the way

**Comment by WillPower3309** [comment]
shadows are very view dependent, maybe we keep them in `render_view` and then have a oneliner for layer surfaces here, since no other processing (other than corner_radius scale) needs to be done:
```c
render_box_shadow(output, output_damage, &shadow_box, config->shadow_color, config->shadow_blur_sigma, corner_radius);
```

We've shown in #150 (sadly) that moving the shadows to render_surface_iterator may not have been what fixed #141

**Comment by ErikReider** [comment]
Layers on the bottom layer shouldn't be able to blur (at least with xray on) due to how the blur optimization currently works. The result is that it blurs itself. (The bottom layer is drawn before the blur optimization and the tiled windows are drawn)

image

**Comment by ErikReider** [comment]
> Moving to render_surface_iterator breaks the smart_corner_radius shadow fixes: [74e7012](https://github.com/WillPower3309/swayfx/commit/74e7012b21d48200cf67608f965befe5a308a8bb)

Why wouldn't we be able to move that logic into `render_surface_iterator`? All of the needed pointers are available for tiled windows (when `view` isn't null). Or did I misunderstand what you wrote? 😅

**Comment by WillPower3309** [comment]
> > Moving to render_surface_iterator breaks the smart_corner_radius shadow fixes: [74e7012](https://github.com/WillPower3309/swayfx/commit/74e7012b21d48200cf67608f965befe5a308a8bb)
> 
> Why wouldn't we be able to move that logic into `render_surface_iterator`? All of the needed pointers are available for tiled windows (when `view` isn't null). Or did I misunderstand what you wrote? 😅

Was just a bug caused by the below line:
```
corner_radius = (con->corner_radius + state->border_thickness) * wlr_output->scale;
```

In render_surface iterator. Changed it to be
```
corner_radius += state->border_thickness
``
In #150 and then applying the scale after.

**Comment by WillPower3309** [comment]
I've messed up a bunch upstream for this branch, I'll rebase it + fix things up tomorrow

**Comment by WillPower3309** [comment]
Will need a hefty rebase with #150 merged

**Comment by ErikReider** [comment]
> Will need a hefty rebase with #155 merged

I'm on it! o7

File: include/sway/desktop/fx_renderer/fx_renderer.h [hunk 6]
```diff
@@ -9,6 +9,19 @@
...
+		.dim = 0.0f, \
+		.dim_color = config->dim_inactive_colors.unfocused, \
+		.corner_radius = 0, \
+		.saturation = 1.0f, \
+		.has_titlebar = false, \
+		.blur = false, \
+		.can_blur_xray = false, \
+		.shadow = false, \
+	}
+
```
Comment (WillPower3309) [inline include/sway/desktop/fx_renderer/fx_renderer.h:24]: I think we should keep this one as a function

File: include/sway/desktop/fx_renderer/fx_renderer.h [hunk 7]
```diff
@@ -33,6 +46,8 @@ struct decoration_data {
 	float *dim_color;
 	bool has_titlebar;
 	bool blur;
+	bool can_blur_xray;
```
Comment (WillPower3309) [inline include/sway/desktop/fx_renderer/fx_renderer.h:49]: We should be able to tell if a layer surface can be x-rayed with it's layer height + if the x-ray setting is enabled, don't think we need this variable

File: sway/commands/blur.c [hunk 8]
```diff
@@ -12,7 +16,7 @@ struct cmd_results *cmd_blur(int argc, char **argv) {
 
 	struct sway_container *con = config->handler_context.container;
 
-	bool result = parse_boolean(argv[0], config->blur_enabled);
+	bool result = cmd_blur_parse_value(argv[0]);
```
Comment (WillPower3309) [inline sway/commands/blur.c:19]: What's the reasoning for this change?

**Review by WillPower3309 — COMMENTED** [review]
Few questions / comments

File: include/sway/desktop/fx_renderer/fx_renderer.h (hunk 7 above)
Comment (ErikReider) [reply include/sway/desktop/fx_renderer/fx_renderer.h:49]: That's true! This was added before we passed the layer pointer to the function 😅

File: sway/commands/blur.c (hunk 8 above)
Comment (WillPower3309) [reply sway/commands/blur.c:19]: Yeah I think we should keep the one liners as is. Good thinking with the corner radius though!

File: sway/desktop/render.c [hunk 9]
```diff
@@ -430,14 +481,37 @@ static void render_surface_iterator(struct sway_output *output,
...
+	render_surface_iterator(output, view, surface, _box, _data);
+
+	struct render_data *data = _data;
+	struct wlr_output *wlr_output = output->wlr_output;
+	struct wlr_box dst_box = *_box;
+	pixman_region32_t *output_damage = data->damage;
+
+	struct decoration_data deco_data = data->deco_data;
+	// TODO: can layers have subsurfaces?
+	bool is_subsurface = view ? view->surface != surface : false;
```
Comment (WillPower3309) [inline sway/desktop/render.c:497]: pretty sure that null will be passed to this function every time, perhaps we should remove view from the function params, and remove this logic

**Comment by WillPower3309** [comment]
Rebased + made some changes so we aren't rendering shadows in the surface iterator, and some minor refactors

**Comment by WillPower3309** [comment]
We do lose xray for layer shell popups here, but if they're really needed we can look into returning that functionality

File: sway/desktop/render.c [hunk 10]
```diff
@@ -430,14 +481,33 @@ static void render_surface_iterator(struct sway_output *output,
...
+		struct wlr_box *_box, void *_data) {
+	// render the layer's surface
+	render_surface_iterator(output, view, surface, _box, _data);
+
+	struct render_data *data = _data;
+	pixman_region32_t *output_damage = data->damage;
+	struct decoration_data deco_data = data->deco_data;
+
+	// render shadow
+	if (deco_data.shadow && config_should_parameters_shadow()) {
```
Comment (WillPower3309) [inline sway/desktop/render.c:496]: TODO: Check if layers can have subsurfaces

Comment (ErikReider) [reply sway/desktop/render.c:498]: Maybe we could add a pointer to the top most surface to `render_data`? We could add another field to `sway_surface_iterator_func_t` but that would make it more "difficult" to rebase from upstream. Any ideas? :)

Comment (WillPower3309) [reply sway/desktop/render.c:498]: I have an idea here, will fix later today

Comment (WillPower3309) [reply sway/desktop/render.c:498]: Can check if it's a subsurface in output.c where we pass the decoration_data. We can just set blur to false there if it's a subsurface!

I'll implement it this evening, but if you're around you're welcome to add it + merge

**Review by WillPower3309 — APPROVED** [review]

File: sway/desktop/output.c [hunk 11]
```diff
@@ -199,6 +199,13 @@ void output_layer_for_each_toplevel_surface(struct sway_output *output,
 	wl_list_for_each(layer_surface, layer_surfaces, link) {
 		struct wlr_layer_surface_v1 *wlr_layer_surface_v1 =
 			layer_surface->layer_surface;
+
+		struct render_data *data = user_data;
+		data->deco_data.blur = layer_surface->layer != ZWLR_LAYER_SHELL_V1_LAYER_BACKGROUND ?
```
Comment (WillPower3309) [inline sway/desktop/output.c:204]: If there's a way to check if a layer is a subsurface (off of my head I don't think there currently is), we should check here

**Review by WillPower3309 — APPROVED** [review]
One thing to look into, but otherwise looks good! I won't have capacity to look into it until this evening

File: sway/desktop/output.c (hunk 11 above)
Comment (ErikReider) [reply sway/desktop/output.c:204]: Might have found a way of doing this. Just noticed that the `wlr_surface_state` struct has a list of surfaces above in the tree, so something like this works :)

```c
bool is_subsurface = wl_list_length(&surface->current.subsurfaces_above) > 0;
```

File: sway/ipc-json.c (hunk 5 above)
Comment (ErikReider) [reply sway/ipc-json.c:360]: Forgot about this, thanks @RicArch97 for the reminder 😅

**Pull request merged**